src/main/scala/scalismo/common/Scalar.scala

```diff
@@ -0,0 +1,283 @@
+      val l = n.toLong
+      l match {
+        case Long.MaxValue => ULong(BigDecimal(n).toLong)
+        case _ => ULong(l)
+      }
+    }
+
+    override def createArray(data: Array[Long]): ValueClassScalarArray[ULong, Long] = ValueClassScalarArray(data)(this)
+    override def toUnderlying(u: ULong): Long = u.toLong
+    override def fromUnderlying(p: Long): ULong = ULong(p)
+  }
+}
+
+sealed trait ScalarArray[S] {
+  def apply(index: Int): S
+  def length: Int
+  final lazy val size = length
+
+  final def isDefinedAt(index: Int): Boolean = index < size && index >= 0
+
+  def map[T: Scalar: ClassTag](f: S => T): ScalarArray[T]
+
+  def iterator: Iterator[S]
+}
+
+abstract case class AbstractScalarArray[S, U](protected[scalismo] val rawData: Array[U]) extends ScalarArray[S] {
+  override final def length: Int = rawData.length
+
+  protected def fromUnderlying(u: U): S
+
+  override def map[T: Scalar: ClassTag](f: S => T): ScalarArray[T] = {
+    val toScalar = implicitly[Scalar[T]]
+    toScalar match {
+      case s: PrimitiveScalar[T] => s.createArray(rawData.map { u => f(fromUnderlying(u)) })
+      case s: ValueClassScalar[T, _] => s.convertArray[U](rawData, { u => f(fromUnderlying(u)) })
+    }
+  }
+
+}
+
+final class PrimitiveScalarArray[S <: AnyVal: ClassTag](rawData: Array[S]) extends AbstractScalarArray[S, S](rawData) {
+
+  override protected def fromUnderlying(u: S): S = u
+
+  override def apply(index: Int): S = rawData(index)
+
+  override def iterator: Iterator[S] = rawData.iterator
+
+  override lazy val hashCode: Int = rawData.deep.hashCode()
+
+  override def canEqual(that: Any): Boolean = that.isInstanceOf[PrimitiveScalarArray[_]]
+
+  override def equals(that: Any) = {
+    that match {
+      case a: PrimitiveScalarArray[_] => (this canEqual that) && this.rawData.deep == a.rawData.deep
+      case _ => false
+    }
+  }
+
+}
+
+final class ValueClassScalarArray[S <: AnyVal, U <: AnyVal](rawData: Array[U])(implicit scalar: ValueClassScalar[S, U]) extends AbstractScalarArray[S, U](rawData) {
+
+  override protected def fromUnderlying(u: U): S = scalar.fromUnderlying(u)
+
+  override def apply(index: Int): S = fromUnderlying(rawData(index))
+
+  override def iterator: Iterator[S] = rawData.iterator.map(scalar.fromUnderlying)
+
+  override lazy val hashCode: Int = rawData.deep.hashCode()
+
+  override def canEqual(that: Any): Boolean = that.isInstanceOf[ValueClassScalarArray[_, _]]
+
+  override def equals(that: Any) = {
+    that match {
+      case a: ValueClassScalarArray[_, _] => (this canEqual that) && this.rawData.deep == a.rawData.deep
+      case _ => false
+    }
+  }
+
+}
+
+object ValueClassScalarArray {
+  def apply[S <: AnyVal, U <: AnyVal](array: Array[U])(implicit s: ValueClassScalar[S, U]): ValueClassScalarArray[S, U] = new ValueClassScalarArray[S, U](array)(s)
+}
+
+object ScalarArray {
+
+  def apply[T: Scalar: ClassTag](array: Array[T]): ScalarArray[T] = {
```

Comment:
Would it make sense to allow for the creation of ScalarArrays from arbitrary traversables? 
+ more scalaish
- it is slower, and hence should be discouraged